DOC-11078: Creates Scopes and Collections Config page and child page in the nav (#740)
couchbaselabs/docs-sync-gateway: modules/ROOT/pages/scopes-and-collections-config.adoc

= Scopes and Collections Configuration for Sync Gateway
:page-aliases: learn/scopes-and-collections-config.adoc
ifdef::show_edition[:page-edition: {release}]
ifdef::prerelease[:page-status: {prerelease}]
:page-role:
:page-content: conceptual

> // :description: pass:q[About Sync Gateway _Roles_ and their part in secure cloud-to-edge enterprise data synchronization.]

Reviewer: Remove comment?

```suggestion
:description: pass:q[How to configure Scopes and Collections for Sync Gateway - Examples with descriptions.]
```

> For a successful configuration of your Couchbase Sync Gateway Scopes and Collections, use this helpful example. 

Reviewer 2: This section explains how to configure Scopes and Collections for Sync Gateway. It describes configuration options and provides example configuration code.

**Understanding buckets, scopes, and collections**
Each database can contain multiple buckets. Each bucket can hold multiple scopes, and each scope can contain multiple collections.

When you create a bucket, a bucket._default._default element is inherently included. (see the documentation on [Default Scopes and Collections](https://github.com/couchbase/docs-sync-gateway/blob/be0913dac8a8f27a2f06554fd759f3ba9448a8a8/modules/ROOT/pages/server/learn/data/scopes-and-collections.adoc#default-scope-and-collection) in Couchbase Server.) Couchbase creates the default bucket to ensure backward compatibility.

Note: Only Couchbase Server (CBS) allows using the word 'default' in naming conventions. As a result,you may see the term _default.default when creating a new bucket.

First, it's important to remember that buckets can hold multiple scopes, and each scope can contain multiple collections.

> When you create a bucket, a `bucket._default._default` element is inherently included, as per the documentation on xref:server/learn/data/scopes-and-collections.adoc#default-scope-and-collection[Default Scopes and Collections] in Couchbase Server.

Reviewer 3: I don't think this needs to be named here, but this is referred to as a `keyspace` in N1QL https://docs.couchbase.com/server/current/n1ql/n1ql-intro/sysinfo.html

Keyspaces in sync gateway are similar except they are `<db>.<scope>.<collection>`

> When you create a bucket, a bucket._default._default element is inherently included, as per the documentation on xref:server/learn/data/scopes-and-collections.adoc#default-scope-and-collection[Default Scopes and Collections] in Couchbase Server

Reviewer: This is hard to parse without some snippet styling. Missing period.
```suggestion
When you create a bucket, a `bucket._default._default` element is inherently included, as per the documentation on xref:server/learn/data/scopes-and-collections.adoc#default-scope-and-collection[Default Scopes and Collections] in Couchbase Server.
```

The default is unique to the bucket creation process and predominantly serves the purpose of backward compatibility.

> Please be aware that only Couchbase Server (CBS) allows using the word 'default' in naming conventions. 

Reviewer 3: ```suggestion
Please be aware that only Couchbase Server (CBS) allows using the word `_default` in naming conventions. 
```

These naming conventions are private to CBS, not accessible to users.

> This explains why you may see the term _default.default when creating a new bucket. 

Reviewer 3: ```suggestion
This explains why you may see the term `_default._default` when creating a new bucket. 
```

Couchbase encourage you to explore and experiment with this concept within your Couchbase instance for practical experience.

== Configuration Options

You have these options to configure Scopes and Collections Sync for Gateway:

* *custom scope/custom collection (the option we cover here)* +
Gain maximum flexibility in organizing and accessing your data by customizing it to suit your needs.

> [Note]

Reviewer: Needs to be uppercase.

```suggestion
[NOTE]
```

====

> Scopes is not required for upgrading to Sync Gateway 3.1. 

Reviewer 3: ```suggestion
The use of collections is not required for upgrading to Sync Gateway 3.1. 
```
We generally refer to things as "collections" even though all collections have a scopes parameter which would be the internal and config file name.

> Scopes is not required for upgrading to Sync Gateway 3.1. 

Reviewer 2: Scopes are not required...

this note seems a little out of place. Move it after the bulleted list?

If it is not specified in the configuration, the previous versions are still compatible with the new version. Therefore, the provided examples are not necessary for upgrading to the latest version of Sync Gateway.
====

* *default scope/custom collection* +

> Create a custom collection with the default scope can lead to efficient data organization and quicker access control management.

Reviewer 2: Create a custom collection with the default scope for efficient data organization and quicker access control managment.

* *default scope/default collection* +

> Work within the default Scopes and Collections. This makes managing data easy and simple without creating extra scopes or collections.

Reviewer 2: Work with the default scopes and collections to make managing data easy without creating extra scopes or collections.

> [Note]

Reviewer: ```suggestion
[NOTE]
```

====

> Importing a dataset created in previous versions of Couchbase Server, the documents will now be saved automatically in the "_default" collection and scope.

Reviewer 3: ```suggestion
Importing a dataset created in previous versions of Couchbase Server, the documents will now be saved automatically in the `_default` scope and `_default` collection.
```

Author: Changed to:

When you import a dataset created in previous versions of Couchbase Server, the documents will automatically be saved in the `_default` scope and `_default` collection.

> Importing a dataset created in previous versions of Couchbase Server, the documents will now be saved automatically in the "_default" collection and scope.

Reviewer 2: When you import a dataset...

====

> == Configuration Example: _custom scope/_custom collection

Reviewer 3: ```suggestion
== Configuration Example: custom scope/custom collection
```

In this example, we cover a configuration with a custom Scope and 2 custom Collections:

[source,javascript]
----
{
  "name": "db",
  "bucket": "bucket",
  "scopes" : {
    "scope1": {
      "collections": {
        "collection1" : {
            "sync": `
                function(doc, oldDoc, meta) {
                    if (doc.type != "default") {
                        throw({forbidden : "Rejected document"})
                    }
                    channel("legacy")
                }
            `,
            "import_filter": `
              function(doc) {
                return doc.type == "mobile"
              }
            `
        },
        "collection2" : {
          "sync": `
            function(doc, oldDoc, meta) {
              channel("collection1")
            }
          `,
          "import_filter": `
              function(doc) {
                return doc.type == "mobile"
              }
          `
        }
      }
    }
  },
  "num_index_replicas": 0
}
----

== Collections Configuration Description

> Here is the description for the custom `Scope` with 2 custom `Collections` configuration. 

Reviewer 2: (Avoid starting a section with Here is, here are)

This section describes the custom scope

== Database Information

=== db
This sets the name of the database to "db".

=== bucket
This indicates the bucket's name where the data will be stored, in this case, "bucket".

== Scopes
A Scope can contain a set of Collections. 
You have the option to use the default provided by Couchbase or customize it to fit your specific needs.

=== scope1

The "scopes" field is not required for upgrading to Sync Gateway 3.1. 
Even if this field is not specified in the configuration, the previous versions of the config file are still compatible with the new version.
Therefore, the provided examples are not essential for upgrading to Sync Gateway's latest version.

== Collections
A Collection can contain a set of documents. 
You have the option to use the default provided by Couchbase or customize it to fit your specific needs.

=== Collection1

=== Sync Function:
[source,javascript]
----
function(doc, oldDoc, meta) {
    if (doc.type != "default") {
        throw({forbidden : "Rejected document"});
    }
    channel("legacy");
}
----
While the example sync function here has a specific purpose, it's crucial to understand that sync functions can perform various actions.

> Whenever a document is synced or updated in collection1, a function called `sync` is executed. 

Reviewer 3: link to https://docs.couchbase.com/sync-gateway/current/sync-function.html ?

Author: added..

This function has three parameters: the new document (doc), the existing document (oldDoc), and document metadata (meta). 
If the type field of the document is not `default`, the function will throw an error message saying "Rejected document" and call the `channel` function with the argument "legacy".

=== Import Filter:
[source,javascript]
----
function(doc) {
    return doc.type == "mobile";
}
----

> When importing documents into collection1, there's an import filter that checks the `type` field value to determine if a document should be imported. 

Reviewer 3: link to https://docs.couchbase.com/sync-gateway/current/import-filter.html ?

Author: Added..

> When importing documents into collection1, there's an import filter that checks the `type` field value to determine if a document should be imported. 

Reviewer 2: When importing documents into collection 1, an import filter checks the 'type' field...

If the `type` field equals `mobile`, the filter returns true, and the document is imported.

== Collections
A Collection can contain a set of documents. 
You have the option to use the default provided by Couchbase or customize it to fit your specific needs.

=== Collection2

=== Sync Function
[source,javascript]
----
function(doc, oldDoc, meta) {
    channel("collection1");
}
----

> Whenever a document is synced or updated in `collection2`, its sync function is triggered. 

Reviewer 3: One point I'm not sure how to address is that the sync function happens any time a document is changed, and sync doesn't have meaning, these are the times it happens:

- a document is updating independently in an a bucket on CBS by a process that is not sync gateway. If this happens and `auto_import` is true, the `import_filter` will decide whether the document is imported. If the document is imported the sync function will be run.
- a document is replicated from a couchbase lite client
- a document is created/modified through the sync gateway REST API

Author: Added(as a note):

The sync function occurs every time a document is modified, regardless of whether or not it has any significance. Here are the examples when it happens:

A document within a CBS bucket is updated by a process other than Sync Gateway. If the auto_import feature is on, the import_filter will decide if the document is eligible for import. If it passes, the sync function will activate.

A document has been copied from a Couchbase Lite client.

A document can be created or modified by using the Sync Gateway REST API.

This function calls the `channel` function with `collection1` as an argument, which means it will communicate with the sync function of `collection1`.

> [Note]

Reviewer: ```suggestion
[NOTE]
```

====
While the example sync function here has a specific purpose, it's crucial to understand that xref:sync-function.adoc[Sync Functions] can perform various actions.
====

=== Import Filter
[source,javascript]
----
function(doc) {
    return doc.type == "mobile";
}
----

When importing documents for `collection2`, the import filter applies the same criteria as `collection1`. 
This filter is based on the value of the `type` field and is used to filter the documents.

> [Note]

Reviewer: ```suggestion
[NOTE]
```
